File: examples/live-dashboard/edge.py
# A live-dashboard edge in Python - the parity of examples/live-dashboard/edge/main.go and edge.ts. It
# holds browser SSE connections and pushes each client only the events of the site it is authorized for.
# The plumbing (SSE, per-connection scoped subscribe, backpressure, drain) is SSEStream; this file's job
# is only authorization - map the request to a site, then hand the connection to the stream.
import os

from aiohttp import web

from aether import SSEStream, _sub_evt, def_edge, run_edge

addr = 7392


def site_from_token(request: web.Request):
    # The demo authorization: a bearer token (query ?token= or Authorization header) maps to a site. A
    # real edge would verify a JWT and take the site from its claims.
    tok = request.query.get("token", "")
    if not tok:
        auth = request.headers.get("Authorization", "")
        tok = auth[len("Bearer "):] if auth.startswith("Bearer ") else ""
    return {"tok-site-1": "site-1", "tok-site-2": "site-2"}.get(tok)


async def run(ctx, stop):
    stream = SSEStream(ctx)

    async def handle_events(request: web.Request):
        site = site_from_token(request)
        if not site:
            return web.Response(status=401, text="unauthorized")
        # The client sees ONLY its site's event subject - NATS never delivers anything else.
        return await stream.serve_client(request, _sub_evt(ctx.app, site))  # aether.<app>.<site>.evt

    app = web.Application()
    app.router.add_get("/events", handle_events)

    runner = web.AppRunner(app)
    await runner.setup()
    site = web.TCPSite(runner, "0.0.0.0", addr)
    await site.start()
    ctx.log.info("live-dashboard edge (py) listening", addr=addr)
    try:
        await stop.wait()
    finally:
        # On drain: end the live SSE connections first (stream.close), then close the server.
        stream.close()
        await runner.cleanup()


if __name__ == "__main__":
    run_edge(def_edge(name=os.environ.get("AETHER_NAME") or "dashboard", run=run))

File: examples/live-dashboard/edge.ts
// A live-dashboard edge in TypeScript - the parity of examples/live-dashboard/edge/main.go. It holds
// browser SSE connections and pushes each client only the events of the site it is authorized for. The
// plumbing (SSE, per-connection scoped subscribe, backpressure, drain) is SSEStream; this file's job is
// only authorization - map the request to a site, then hand the connection to the stream.
import { startEdge, SSEStream, subjects } from "@hamicek/aether";
import * as http from "node:http";

const addr = 7392;

// siteFromToken is the demo authorization: a bearer token (query ?token= or Authorization header) maps to
// a site. A real edge would verify a JWT and take the site from its claims.
function siteFromToken(req: http.IncomingMessage): string | null {
  const url = new URL(req.url ?? "/", "http://localhost");
  let tok = url.searchParams.get("token") ?? "";
  if (!tok) tok = (req.headers["authorization"] ?? "").replace(/^Bearer /, "");
  if (tok === "tok-site-1") return "site-1";
  if (tok === "tok-site-2") return "site-2";
  return null;
}

let server: http.Server;
let stream: SSEStream;

await startEdge({
  init: (ctx) => {
    stream = new SSEStream(ctx);
    server = http.createServer(async (req, res) => {
      const url = new URL(req.url ?? "/", "http://localhost");
      if (url.pathname !== "/events") {
        res.writeHead(404);
        res.end("no route");
        return;
      }
      const site = siteFromToken(req);
      if (!site) {
        res.writeHead(401);
        res.end("unauthorized");
        return;
      }
      // The client sees ONLY its site's event subject - NATS never delivers anything else.
      await stream.serveClient(req, res, subjects.eventLog(ctx.app, site)); // aether.<app>.<site>.evt
    });
    ctx.log.info("live-dashboard edge (ts) listening", { addr });
  },

  // On drain: end the live SSE connections first (stream.close), then close the server.
  run: (ctx, stop) =>
    new Promise<void>((resolve, reject) => {
      server.on("error", reject);
      server.listen(addr);
      void stop.then(() => {
        stream.close();
        server.close(() => resolve());
      });
    }),
});
